docs/tech_specs/screens_by_alert/tech_spec.md

```diff
@@ -0,0 +1,187 @@
+# Screens by alert real-time data source - tech spec
+
+## Need
+
+Screenplay (and maybe other services, in the future) needs real-time insight into which screens are
+displaying info about a given service alert.
+
+## High-level architecture
+
+To store and provide this new data, we will add a new caching layer to the Screens application.
+
+To keep the cache in one shared location, we will use memcached rather than an ETS table or GenServer/other stateful process running on each server instance. This follows the precedent of the V3 API, which uses memcached to keep track of per-user request counts and apply throttling appropriately no matter how many instances it's scaled to.
+
+### Data structure
+
+Cached values will be as follows, defined as Elixir typespecs:
+```ex
+@type alert_id :: String.t()
+@type screen_id :: String.t()
+@type timestamp :: integer()  # we'll use System.monotonic_time/0 to produce these
+@type timestamped_screen_id :: {screen_id, timestamp}
+
+@type cache :: %{
+  "screens_by_alert." <> alert_id => list(timestamped_screen_id),
+  # metadata to make the "self-refresh" mechanism possible
+  "screens_last_updated." <> screen_id => timestamp
+}
+```
+
+TTL for `screens_by_alert.*` cache keys will be a number just over the "self-refresh" interval, so that in the case of a screen client going down we don't inaccurately report it as showing no alerts for a short period before the "self-refresh" mechanism kicks in.
```

Comment:
`TTL for 'screens_by_alert.*' cache keys will be a number just over the "self-refresh" interval`: could you explain this idea in different words? I can't get these terms to make sense in my mind. In the typespec for the cache, it seems that the "cache key" for `screens_by_alert.*` is literally that string plus the alert id. How is TTL attached to that?

Reply:
Thanks for the question, I could have been clearer on this!

In this case, **TTL will be handled internally by memcached** and isn't explicitly part of the data structure.

Every time you set or update a key's value in memcached, you can also tell it what that key/value pair's TTL is. See, for example, the `set` [function](https://hexdocs.pm/memcachex/Memcache.html#set/4) made available by the `memcachex` library—it accepts a `:ttl` option to specify TTL. Once that specified amount of time elapses, the key+value will automatically expire from the cache.

Our case is a little tricky though, because we also need to implement some separate TTLs that we handle in our own code—memcached's TTL feature only applies to whole key/value pairs, but we need to also apply TTLs to individual values under each alert ID key.

Assuming we use 60 seconds as the "self-refresh" interval:
```ex
# An example key/value pair in the cache.
# When we set/update this key, we set its TTL to 70 seconds.
# If this key's value isn't updated again within 70 seconds, it will automatically be removed by memcached.
"screens_by_alert.123" => [
  {"PRE-101", 100},
  # ^ One of the values in the list stored under this key.
  # The first tuple element is a screen ID, the second is a timestamp produced by `System.monotonic_time()`.
  # Every time we update this list, we remove any element whose timestamp is from more than 70 seconds ago.
  {"EIG-123", 200},
  {"BUS-456", 300}
]
```

Reply:
I understand the first 3 paragraphs. ✅ 

Half of the 4th paragraph makes sense -- it pairs with your comment to my other question. We have a TTL we provide to memcached, and that time limit will immediately expire a key-value after it has had the opportunity to run the self-refresh code. The definition of "self-refresh" is kinda buried in this doc, but in essence: for each screen ID, we'll look at the cache to see if the screen has recently updated the cache itself (indicated by a very recent timestamp) and if the screen has NOT updated itself (indicating a screen outage), we trigger a run of the widget framework for that screen to force an update of the cache with realtime alert status. 

So, after this self-refresh has had a chance to occur, we'll see one of two things: either the screen affirmed that it still has an alert, or it claims it doesn't have that alert anymore. If the screen still has the alert, the key-value stored in memcached will have been updated (restarting the TTL clock and the self-refresh clock). If the screen no longer has that alert, the key-value timestamp (within memcached) will be unchanged, and then the TTL will expire that key-value.

How is this translating so far?

Reply:
> If the screen still has the alert, the key-value stored in memcached will have been updated (restarting the TTL clock and the self-refresh clock).

Correct

> If the screen no longer has that alert, the key-value timestamp (within memcached) will be unchanged, and then the TTL will expire that key-value.

This is also correct, but I just want to clarify two things:
1. If the screen no longer has that alert, then the cache key for that alert won't get updated. However, other screens might be showing it and thus updating the cache key, so it won't necessarily expire just because this one update no longer included it.
2. The self-refresh clock for this screen will also be updated in this case, just like in the first case where the screen still has the alert. Every screen data update that gets processed (whether from an external request, or from the self-refresh mechanism) will cause the `screens_last_updated.<id>` key for that screen to get updated with a new timestamp value.

Reply:
Thanks! Here's an updated translation, picking up halfway through my last translation, with new questions:

> If the screen no longer has that alert...

... then the best that this screen can do is say “I have these alerts: []”… Can we somehow compare that list of alerts with the cache to more quickly break the relationship between the alert and screen?

If not, then the alert will still seem to be on that screen while other stuff happens…

If other screens still have the alert, they will one-by-one affirm that they should still be in the screens array for that alert in the cache. Plus, they will also be updating the key-value timestamp for that alert in memcached. If this is the case… how does the screen-without-the-alert remove itself from its place in the cached array?

On the other hand, if other screens one by one indicate they, too, no longer have the alert, the key-value will get staler and staler until it’s clear that the alert has been removed from all screens and should be expired. Then the TTL in memcached will kick in to expire that key-value pair.

Reply:
> Can we somehow compare that list of alerts with the cache to more quickly break the relationship between the alert and screen?

Nope, not without having a lot more cache operations per update! That would essentially require us to turn it back into keeping a full representation of a bipartite graph in the cache, with the associated increase in potential race conditions/excessive retries of cache updates. The way we avoid this extra work is by using TTLs and never comparing a previous list of reported alerts to the new list of reported alerts.

> how does the screen-without-the-alert remove itself from its place in the cached array?
Whenever we make any update to the list under an alert key, we first get the existing list and remove any screen IDs that have expired (based on the timestamps they're paired with).

Last paragraph is correct. If all screens stop showing an alert (probably because it's been cleared in alerts-ui) then it and its associated list of screen IDs eventually expires from the cache entirely. Screenplay won't need to wait for the cached value to expire though, if the alert was cleared, because it also checks alert data in the V3 API. It will know that the alert is gone, and simply stop asking for info about it, before the cached data expires.